onnxruntime/test/onnx/gen_test_models.py

```diff
@@ -20,7 +20,7 @@ def parse_arguments():
 
 def write_config(model_dir):
     with open(os.path.join(model_dir, "config.txt"), "w") as f:
-        f.write("per_sample_tolerance:1e-6\n")
+        f.write("per_sample_tolerance:1e-3\n")
```

Comment:
@tracysh How much variance is reasonable/expected from the change to use Eigen::squaredNorm and gemm? A Linux build is failing due to the below diff. I can 'fix' that by making the relative tolerance 1e-2 (in addition to reducing the absolute tolerance to 1e-3), but I'm wondering if this is getting outside the level of floating point accuracy expected. 

Build failure:
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=129310&view=logs&j=efb7335f-9de0-576e-0860-2a077aa78ef7&t=fe3fac92-10f4-5688-eb8f-58f0bf9e8d21

Diff:
2020-04-03 09:47:03.173452783 [E:onnxruntime:Default, runner.cc:444 RunTaskImpl] cdist_float32_euclidean_1000_2000_1:output=C:expected 0.00102735 (3a86a800), got 0 (0), diff: 0.00102735, tol=0.001. 1 of 2000000 differ

Reply:
I changed the unit tests to generate deterministic data to try and investigate, but now the tests pass. I did look into some of the diffs with the local data I have, and if the two numbers are close enough you can get a 0.0 depending on the order of operations. 

Below is python output for two of the test values

a=-1.6629804372787476
b=-1.6622027158737183
a - b = -0.0007777214

a*a + b*b -2*(a*b) = 0.0
a*a + b*b - (2*a*b) = 2.0339834350124875e-07
a*a -(2*a*b) + b*b = 4.4181692260281125e-07
a**2 + b**2 -2*(a*b) = 4.014522403394949e-07
((a-b)**2) = 6.048505838407436e-07

Reply:
Is re-ordering to do the gemm with beta of 0 first and adding with the sum squared values from A and B after more precise? Values from C++ below for the 'a' and 'b' above.

sum squared for A: 2.76550388
sum squared for B: 2.76291776
-2*a*b from gemm:-5.52842140

-2*a*b + ss_a + ss_b: 2.38418579e-07
vs. current of 
ss_a + ss_b + -2*a*b : 0.00000000

Reply:
The python values vary wildly due to implicit promotion for some operations to float64 from float32 (-2*a and a**2 get promoted, a*a or a*b does not). Forcing python to stay in float32 results in it producing the same value for -2*a*b + a*a + b*b as the C++ code.

If the points being compared are close, 2*a*b will be very similar to (a*a + b*b) and subtracting two floating point numbers that are close can be problematic, so subtracting 2*a*b from a*a first and adding b*b second leads to less loss of precision.